language-plutus-core/prop-test/Language/PlutusCore/Gen/Type.hs

```diff
@@ -0,0 +1,252 @@
+{-|
+Module: Language.PlutusCore.Gen.Type
+
+This file contains:
+1. A duplicate of the Plutus Core Abstract Syntax (types and terms)
+2. A kind checker and a type checker
+3. Reduction semantics for types
+-}
+
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE DeriveDataTypeable  #-}
+{-# LANGUAGE FlexibleInstances   #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell     #-}
+
+module Language.PlutusCore.Gen.Type
+  ( TypeBuiltinG (..)
+  , TypeG (..)
+  , ClosedTypeG
+  , toClosedType
+  , checkClosedTypeG
+  , normalizeTypeG
+  ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Gen.Common
+import           Control.Enumerable
+import           Data.Coolean
+import qualified Data.Text as Text
+import           Text.Printf
+
+-- * Enumerating builtin types
+
+data TypeBuiltinG
+  = TyByteStringG
+  | TyIntegerG
+  | TyStringG
+  deriving (Typeable, Eq, Show)
+
+$(deriveEnumerable ''TypeBuiltinG)
+
+-- |Convert generated builtin types to Plutus builtin types.
+toTypeBuiltin :: TypeBuiltinG -> Some (TypeIn DefaultUni)
+toTypeBuiltin TyByteStringG = Some (TypeIn DefaultUniByteString)
+toTypeBuiltin TyIntegerG    = Some (TypeIn DefaultUniInteger)
+toTypeBuiltin TyStringG     = Some (TypeIn DefaultUniString)
+
+
+-- * Enumerating types
+
+data TypeG n
+  = TyVarG n
+  | TyFunG (TypeG n) (TypeG n)
+  | TyIFixG (TypeG n) (Kind ()) (TypeG n)
+  | TyForallG (Kind ()) (TypeG (S n))
+  | TyBuiltinG TypeBuiltinG
+  | TyLamG (TypeG (S n))
+  | TyAppG (TypeG n) (TypeG n) (Kind ())
+  deriving (Typeable, Eq, Show)
+
+$(deriveEnumerable ''TypeG)
+
+
+-- |Convert well-kinded generated types to Plutus types.
+--
+-- NOTE: Passes an explicit `TyNameState`, instead of using a State monad,
+--       as the type of the `TyNameState` changes throughout the computation.
+--       Alternatively, this could be written using an indexed State monad.
+--
+-- NOTE: The error types in Language.PlutusCore.Error are closed, and hence it
```

Comment:
Also, I don't buy this. You can perfectly well define your own local error type, use `MonadErrror` or whatever, and match on it later if it's a problem.

Especially given the note about there maybe being problems due to partiality, we might as well get rid of it!

Reply:
I have had a go at this now.